### PR TITLE
split up CUDA-suffixed dependencies in dependencies.yaml

### PR DESCRIPTION
**Pull request by jameslamb**
## Description

Contributes to https://github.com/rapidsai/build-planning/issues/31

In short, RAPIDS DLFW builds want to produce wheels with unsuffixed dependencies, e.g. `cudf` depending on `rmm`, not `rmm-cu12`.

This PR is part of a series across all of RAPIDS to try to support that type of build by setting up CUDA-suffixed and CUDA-unsuffixed dependency lists in `dependencies.yaml`.

For more details, see:

* https://github.com/rapidsai/build-planning/issues/31#issuecomment-2245815818
* https://github.com/rapidsai/cudf/pull/16183

### Why target 24.08?

This is targeting 24.08 because:

1. it should be very low-risk
2. getting these changes into 24.08 prevents the need to carry around patches for every library in DLFW builds using RAPIDS 24.08

### CODE REVIEW

**Comment by alexbarghi-nv** [comment]
I think we need this applied to rapidsai/cugraph-gnn as well

**Comment by jameslamb** [comment]
Sure @alexbarghi-nv I can do that! You're right, I'd missed that because `cugraph-gnn` wasn't in the initial list on https://github.com/rapidsai/build-planning/issues/31. I'll put up a PR with changes like this there.

**Comment by jameslamb** [comment]
I've updated this based on the suggestions from https://github.com/rapidsai/cudf/pull/16183.

Ran `ci/release/update-version.sh '24.10.00'` and did not see any other places that need updates.

**Comment by jameslamb** [comment]
@alexbarghi-nv I created the following in `cugraph-gnn`:

* PR with changes like this one: https://github.com/rapidsai/cugraph-gnn/pull/16
* issue documenting the work of switching to `rapids-build-backend`: https://github.com/rapidsai/cugraph-gnn/issues/15

File: dependencies.yaml [hunk 1]
```diff
@@ -493,7 +493,7 @@ dependencies:
           - requests
           - nccl>=2.9.9
           - ucx-proc=*=gpu
-          - &ucx_py ucx-py==0.39.*,>=0.0.0a0
+          - &ucx_py_unsuffixed ucx-py==0.39.*,>=0.0.0a0
```
Comment (bdice) [inline dependencies.yaml:496]: Following from my other review on the cuDF PR: I don't care if we use `_unsuffixed` or not. `ucx_py` is fine, just so long as it doesn't say `_conda` if it's used in non-conda lists.

File: dependencies.yaml [hunk 2]
```diff
@@ -850,19 +886,26 @@ dependencies:
...
+          - {matrix: null, packages: [*pylibcugraphops_unsuffixed]}
 
   depends_on_cupy:
     common:
       - output_types: conda
         packages:
           - cupy>=12.0.0
+    # NOTE: this is not broken into groups by a 'cuda_suffixed' selector like
+    #       other packages with -cu{nn}x suffixes in this file because devcontainers-based
+    #       builds that build it from source expect a package with a `-cuda{nn}x` suffix
```
Comment (bdice) [inline dependencies.yaml:908]: > that build it from source

The antecedent is a little unclear here. "it" refers to cugraph, not cupy here. We never build cupy from source in RAPIDS. Maybe we can simplify the comment like:

```suggestion
    # cupy is not built from source in devcontainers, so cuda_suffixed isn't needed
```

Comment (jameslamb) [reply dependencies.yaml:908]: My choice to say "devcontainers-based builds" was a bad one. It's very unclear.

I should have said "DLFW builds".

We do not build `cupy` from source in RAPIDS devcontainers, but we DO build it from source in the RAPIDS DLFW builds. And there, pass the following through to `pip wheel`.

```text
--build-option="--cupy-package-name=cupy-cuda$(cut -d'.' -f1 <<< "${CUDA_VERSION}")x"
```

Here's where `cupy` exposes that option in its configuration:

https://github.com/cupy/cupy/blob/9381230cc55a5bf96bf838a6c6cc077b38fe68d1/install/cupy_builder/_context.py#L81-L83

So it's important, in the DFLW context, that `cugraph` depend on e.g. `"cupy-cuda12x"`, not `"cupy"`.

Maybe we should change that in DLFW in the future, or maybe the `cupy` dependency should be broken out into a separate block called like `depends_on_cupy` here in `dependencies.yaml`, so that these `cuda_suffixed: "false"` entries could be removed.

But with code freeze for 24.08 so close, I wasn't looking to change any behavior... just to reflect the current state in the smallest, least controversial changeset possible to get these merged in.

Given all that... would you support leaving this as-is and instead having a comment like the one I'm proposing on `cuml`'s `cupy` dependency over in https://github.com/rapidsai/cuml/pull/5974?

```text
# NOTE: cupy still has a "-cuda12x" suffix here, because it's suffixed
#       in DLFW builds
```

File: dependencies.yaml (hunk 1 above)
Comment (jameslamb) [reply dependencies.yaml:496]: I'd like to keep the cases like this, I think it's a useful convention that makes the intent slightly clearer.

**Comment by jameslamb** [comment]
Cancelling CI here (and will on other commits in the next 24 hours) to save some CI capacity for other projects that are about to enter code freeze.

File: dependencies.yaml (hunk 2 above)
Comment (bdice) [reply dependencies.yaml:908]: I think we are on the same page with everything but the wording of this comment -- the behavior here is what I expect, but I don't think the comment needs to describe _only_ DLFW.

To the best of my knowledge, there is no such pip package called `cupy` (public or internal/DLFW), so there's _never_ a case where we want cupy to be unsuffixed in `requirements` or `pyproject` lists. `cupy` is only ever used as the conda package name.

Note that for RAPIDS, devcontainers (including outside of DLFW) need the unsuffixed name [because that matches the exclusion list](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L177-L186).

So either way, nothing here is DLFW-specific, if I understand correctly. We should be able to reword this to not mention DLFW and just say devcontainers. Perhaps the "from source" piece of my previous suggestion is irrelevant.

Comment (bdice) [reply dependencies.yaml:908]: Think about it and merge whatever you think fits best for the comment. The behavior is what I expect, so I don't want to get stuck on the comment.

Comment (jameslamb) [reply dependencies.yaml:908]: ok thanks!

I was thinking `cugraph` went into code freeze today, but I realize that wasn't correct. So we do have a bit more time with this one to make it look the way we long-term want it to. I'll push a change here (tomorrow) with a separated-out `depends_on_cupy` list and a less-confusing comment.

Comment (jameslamb) [reply dependencies.yaml:908]: > need the unsuffixed name [because that matches the exclusion list](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L177-L186).

And just to note... the exclusion list there includes both the suffixed and unsuffixed versions. That's what this line there does:

```bash
pip_noinstall+=("${pkg}" "${pkg}-cu.*");
```

Comment (trxcllnt) [reply dependencies.yaml:908]: To clarify, cupy is not captured by that exclusion list -- that list is constructed from RAPIDS packages (things in `manifest.yaml`). cupy is handled in a special case [here](https://github.com/rapidsai/devcontainers/blob/541895260fa9d3d772989c55c64a3ab98323f1c4/features/src/rapids-build-utils/opt/rapids-build-utils/bin/make-pip-dependencies.sh#L197), where we transform it into:
`cupy-cuda[0-9]+x` -> `cupy-cuda${cuda_version_major}x`

This is to work around issues in previous releases' `dependencies.yaml` that always output `cupy-cuda11x` instead of outputting the correct version based on the `cuda` matrix entry.

Comment (jameslamb) [reply dependencies.yaml:908]: Thanks for clarifying that, appreciated!

Comment (jameslamb) [reply dependencies.yaml]: I just pushed, updating the comment to this:

```yaml
# NOTE: This is intentionally not broken into groups by a 'cuda_suffixed' selector like
#       other packages with -cu{nn}x suffixes in this file.
#       All RAPIDS wheel builds (including in devcontainers) expect cupy to be suffixed.
```

Hopefully that more clearly captures the intent.

> *...maybe the `cupy`dependency should be broken out into a separate block called like depends_on_cupy here in `dependencies.yaml`*

🤦🏻 that's already how it works here in `cugraph`, sorry... too many very-similar-looking PRs yesterday.

**Review by bdice — APPROVED** [review]

**Comment by jameslamb** [comment]
conda tests here are failing like this:

```text
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/raft_dask/common/comms.py:28: in <module>
    from .comms_utils import (
E   ImportError: /opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/raft_dask/common/comms_utils.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZN4ucxx6Worker19registerGenericPostESt8functionIFvvEE
```

<details><summary>full stacktrace (click me)</summary>

```text
ImportError while loading conftest '/__w/cugraph/cugraph/python/cugraph-pyg/cugraph_pyg/tests/conftest.py'.
tests/conftest.py:20: in <module>
    from cugraph.dask.comms import comms as Comms
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/__init__.py:14: in <module>
    from cugraph.community import (
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/community/__init__.py:14: in <module>
    from cugraph.community.louvain import louvain
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/community/louvain.py:15: in <module>
    from cugraph.structure import Graph
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/structure/__init__.py:14: in <module>
    from cugraph.structure.graph_classes import (
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/structure/graph_classes.py:15: in <module>
    from .graph_implementation import (
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/structure/graph_implementation/__init__.py:14: in <module>
    from .simpleGraph import simpleGraphImpl
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/structure/graph_implementation/simpleGraph.py:14: in <module>
    from cugraph.structure import graph_primtypes_wrapper
graph_primtypes_wrapper.pyx:26: in init cugraph.structure.graph_primtypes_wrapper
    ???
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/dask/__init__.py:16: in <module>
    from .link_analysis.pagerank import pagerank
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/dask/link_analysis/pagerank.py:30: in <module>
    import cugraph.dask.comms.comms as Comms
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/cugraph/dask/comms/comms.py:19: in <module>
    from raft_dask.common.comms import Comms as raftComms
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/raft_dask/common/__init__.py:16: in <module>
    from .comms import Comms, local_handle
/opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/raft_dask/common/comms.py:28: in <module>
    from .comms_utils import (
E   ImportError: /opt/conda/envs/test_cugraph_pyg/lib/python3.9/site-packages/raft_dask/common/comms_utils.cpython-39-x86_64-linux-gnu.so: undefined symbol: _ZN4ucxx6Worker19registerGenericPostESt8functionIFvvEE
```

</details>

([build link](https://github.com/rapidsai/cugraph/actions/runs/10094936537/job/27922204394?pr=4552))

Manually retriggering all jobs so builds will be done against latest `ucxx` / `raft` nightlies, let's see if that helps.

**Comment by alexbarghi-nv** [comment]
@jameslamb this is a known issue.  I think @nv-rliu is looking into it.

**Comment by jameslamb** [comment]
> @jameslamb this is a known issue. I think @nv-rliu is looking into it.

Ah ok thanks @alexbarghi-nv ! Then I'll cancel the CI run here, to give back some capacity while it's being worked on.

@nv-rliu if you have an issue or PR, could you link it here so I can follow along?

**Comment by jameslamb** [comment]
restarted all CI here, to hopefully pick up new artifacts with the fixes from https://github.com/rapidsai/raft/pull/2394 and https://github.com/rapidsai/raft/pull/2398

**Comment by jameslamb** [comment]
I've cancelled the 3 `conda-python-tests` jobs, to release some CI capacity back to the rest of RAPIDS.

These have been stuck for about 2 hours. There's some issue causing 1 test to not complete (unrelated to this PR), and it's being temporarily skipped in #4559.
